### PR TITLE
[BEAM-7103] Adding AvroCoderTranslator for cross-language data transfer

### PR DESCRIPTION
**Pull request by ihji**
Adding Java AvroCoderTranslator and Python AvroCoder for a simple dict type cross-language data transfer.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ihji** [comment]
R: @chamikaramj

**Comment by ihji** [comment]
Run JavaPortabilityApi PreCommit

**Comment by ihji** [comment]
+ @mxm

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AvroGenericCoderTranslator.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+public class AvroGenericCoderTranslator implements CoderTranslator<AvroGenericCoder> {
+  @Override
+  public List<? extends Coder<?>> getComponents(AvroGenericCoder from) {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public byte[] getPayload(AvroGenericCoder from) {
+    try {
+      return from.getSchema().toString().getBytes("UTF-8");
```
Comment (mxm) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AvroGenericCoderTranslator.java]:
```suggestion
      return from.getSchema().toString().getBytes(Charsets.UTF_8);
```

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AvroGenericCoderTranslator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException("failed to encode schema.");
+    }
+  }
+
+  @Override
+  public AvroGenericCoder fromComponents(List<Coder<?>> components, byte[] payload) {
+    Schema schema;
+    try {
+      schema = new Schema.Parser().parse(new String(payload, "UTF-8"));
```
Comment (mxm) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AvroGenericCoderTranslator.java]:
```suggestion
      schema = new Schema.Parser().parse(new String(payload, Charsets.UTF_8));
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/AvroGenericCoderTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+          .name("age")
+          .type()
+          .intType()
+          .intDefault(0)
+          .name("hometown")
+          .type()
+          .nullable()
+          .stringType()
+          .noDefault()
+          .endRecord();
```
Comment (mxm) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/AvroGenericCoderTest.java]: Could be interesting to add a custom type, e.g. a PoJo.

File: sdks/python/apache_beam/coders/avro_generic_coder.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+  def __init__(self, schema):
+    self.schema = schema
+
+  def _create_impl(self):
+    return AvroGenericCoderImpl(self.schema)
+
+  def is_deterministic(self):
+    # TODO: need to confirm if it's deterministic
+    return False
```
Comment (mxm) [inline sdks/python/apache_beam/coders/avro_generic_coder.py]: You have probably seen `AvroDeterminismChecker`. So false seems safe here if we do not have a checker in Python.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroGenericCoder.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.coders;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+/** GenericRecord Avro Coder for simple dict type cross-language data transfer. */
+public class AvroGenericCoder extends AvroCoder<GenericRecord> {
+  protected AvroGenericCoder(Schema schema) {
```
Comment (chamikaramj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroGenericCoder.java]: Also, should we be adding this to standard coder definitions in [1] and [2].

[1] https://github.com/apache/beam/blob/master/model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml
[2] https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/beam_runner_api.proto#L557

File: sdks/python/apache_beam/coders/avro_generic_coder.py (hunk 4 above)
Comment (ihji) [reply sdks/python/apache_beam/coders/avro_generic_coder.py]: https://issues.apache.org/jira/browse/BEAM-7308

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroGenericCoder.java (hunk 5 above)
Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroGenericCoder.java]: I think we can follow this up in a separate issue.
https://issues.apache.org/jira/browse/BEAM-7309

**Comment by ihji** [comment]
Figured out that we can just use AvroCoder instead of creating AvroGenericCoder in Java SDK.

**Comment by chamikaramj** [comment]
Thanks. LGTM.

**Pull request merged**